src/fparser/sourceinfo.py

```diff
@@ -196,12 +208,25 @@ def get_source_info(file_candidate):
             return FortranFormat(True, True)
 
     if hasattr(file_candidate, 'read'):
+        # If the candidate object has a "read" method we assume it's a file
+        # object.
+        #
+        # If it is a file object then it may be in the process of being read.
+        # As such we need to take a note of the current state of the file
+        # pointer so we can restore it when we've finished what we're doing.
+        #
         pointer = file_candidate.tell()
         file_candidate.seek(0)
         source_info = get_source_info_str(file_candidate.read())
         file_candidate.seek(pointer)
         return source_info
     else:
+        # No "read" method so presumably it's a filename string. In which case
```

Comment:
Please could you firm-up on the "presumably" by checking that it is a string and raising an appropriate error if not. (This will then require a further test - apologies.)